sklearn/utils/_openmp_helpers.pyx

```diff
@@ -59,4 +59,13 @@ cpdef _openmp_effective_n_threads(n_threads=None):
         # OpenMP disabled at build-time => sequential mode
         return 1
 
-    
+
+cdef inline int _openmp_thread_num() nogil:
```

Comment:
Indeed, I had probably that file folded in the github diff view when I wrote this comment and couldn't find it with ctrl-f.

Are there legitimate cases where the requested `num_threads` is not the actual `_openmp_thread_num()`?

Reply:
On the first hand, `num_threads` represents the total number of threads used for reductions via `parallel`-context manager. Depending on the load, it is less than or equal to the effective number of threads returned by `_openmp_num_threads`; see [its definition](https://github.com/scikit-learn/scikit-learn/blob/ca942a5a9ce513f0b235d21003ef99de0700e19d/sklearn/metrics/_pairwise_distances_reduction.pyx#L258).

On the other hand, `thread_num` which is returned by `_openmp_thread_num` represents the number (identifier per say) of the thread. This is the value used in template methods' scope to chosen the proper datastructures.

---

Note that `parallel`-context manager uses the effective number of threads when it makes sense, for instance [when sorting datastructures post-reduction](https://github.com/scikit-learn/scikit-learn/blob/ca942a5a9ce513f0b235d21003ef99de0700e19d/sklearn/metrics/_pairwise_distances_reduction.pyx#L621).

Reply:
Does this answer your question, @ogrisel?

Reply:
@ogrisel this is the thread number (not the number of threads). This utility functions is just a wrapper around omp_get_thread_num in case we compile without openmp. It's only purpose is to avoid repeating the IF/ELSE each time we need to call omp_get_thread_num